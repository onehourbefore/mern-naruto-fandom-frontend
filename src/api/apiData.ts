export const hostName = 'http://localhost:5000'