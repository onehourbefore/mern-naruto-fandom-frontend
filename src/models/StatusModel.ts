export enum StatusModel {
    LOADING = 'loading',
    ERROR = 'error',
    SUCCESS = 'success',
    IDLE = 'idle'
}