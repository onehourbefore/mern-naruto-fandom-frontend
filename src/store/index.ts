import { configureStore, combineReducers } from '@reduxjs/toolkit';
import { useDispatch, TypedUseSelectorHook, useSelector } from 'react-redux';

import posts from './slices/postsSlice';
import filter from './slices/filterSlice';
import onePost from './slices/onePostSlice';
import registration from './slices/registrationSlice';
import authorization from './slices/authorizationSlice';
import experts from './slices/expertSlice';
import comments from './slices/commentSlice';
import editForm from './slices/editFormSlice';

export const rootReducer = combineReducers ({
    posts,
    filter,
    onePost,
    registration,
    authorization,
    experts,
    comments,
    editForm,
});

export const store = configureStore ({
    reducer: rootReducer,
    middleware: getDefaultMiddleware => getDefaultMiddleware (),
    devTools: process.env.NODE_ENV !== 'production'
});

export type RootState = ReturnType <typeof store.getState>;
export const useAppDispatch: () => typeof store.dispatch = useDispatch;
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;