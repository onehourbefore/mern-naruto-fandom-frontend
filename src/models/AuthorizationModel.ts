import { StatusModel } from "./StatusModel"

export type UserDataType = {
    id: string,
    email: string,
    name: string,
    role: string,
    avatar: string,
    created: number,
    liked: string [] | []
}

export type ResponseType = {
    accessToken: string,
    refreshToken: string,
    user: UserDataType
}

export interface AuthorizationState {
    user: UserDataType,
    status: StatusModel,
    deleteProfileStatus: StatusModel
}