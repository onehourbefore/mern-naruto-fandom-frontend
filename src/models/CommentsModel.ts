import { StatusModel } from "./StatusModel"

export type CommentType = {
    _id: string,
    postID: string,
    date: string,
    name: string,
    email: string,
    body: string,
    avatar: string,
}

export interface CommentsState {
    comments: [] | CommentType [],
    status: StatusModel,
}

export type CommentProps = {
    data: {
        _id: string,
        name: string,
        email: string,
        body: string,
        avatar: string,
        date: string,
    },
    postID: string | undefined,
    commentIndex: number,
}

export type CommentFormProps = {
    setVisible: (arg: boolean) => void,
}

export type CommentFromSidebarProps = {
    comment: CommentType,
}