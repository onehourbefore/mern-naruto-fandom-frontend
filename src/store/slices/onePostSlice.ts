import { createSlice, createAsyncThunk, PayloadAction, isRejectedWithValue } from '@reduxjs/toolkit';
import $api from '../../http';
import { PostType, OnePostState } from '../../models/PostModel';
import { CommentType } from '../../models/CommentsModel';
import { StatusModel } from '../../models/StatusModel';
import { OnePostThunks } from './thunks/OnePostThunks';

const onePostThunks = new OnePostThunks();

const initialState: OnePostState = {
    active: { post: null, status: StatusModel.IDLE },
    likeStatus: StatusModel.IDLE,
    commentStatus: StatusModel.IDLE,
};

export const fetchOnePost = createAsyncThunk(
    'fetchOnePost',
    async(id: string, { rejectWithValue }) => {
        try {
            return await onePostThunks.fetchOnePostCallback(id);
        } catch(e: any) {
            return rejectWithValue(e.message);
        }
    }
);

export const addLike = createAsyncThunk(
    'addLike',
    async(postID: string, { rejectWithValue }) => {
        try {
            return await onePostThunks.addLikeCallback(postID);
        } catch(e: any) {
            return rejectWithValue(e.message);
        }
    }
);

export const removeLike = createAsyncThunk(
    'removeLike',
    async(postID: string, { rejectWithValue }) => {
        try {
            return await onePostThunks.removeLikeCallback(postID);
        } catch(e: any) {
            return rejectWithValue(e.message);
        }
    }
);

export const addComment = createAsyncThunk(
    'addComment',
    async(query: {postID: string, body: string}, { rejectWithValue }) => {
        try {
            return await onePostThunks.addCommentCallback(query);
        } catch(e: any) {
            return rejectWithValue(e.message);
        }
    }
);

export const removeComment = createAsyncThunk(
    'removeComment', 
    async(query: { postID: string, commentID: string }, { rejectWithValue }) => {
        try {
            return await onePostThunks.removeCommentCallback(query);
        } catch(e: any) {
            return rejectWithValue(e.message);
        }
    }
);

export const onePostSlice = createSlice({
    name: 'onePost',
    initialState,
    reducers: {
        clearActivePost: (state) => {
            state.active.post = null
        }
    },
    extraReducers: (builder) => {
        builder
        .addCase(fetchOnePost.pending, (state) => {
            state.active.status = StatusModel.LOADING;
        })
        .addCase(fetchOnePost.fulfilled, (state, action: PayloadAction <PostType>) => {
            state.active.post = { ...action.payload };
            state.active.status = StatusModel.SUCCESS;
        })
        .addCase(fetchOnePost.rejected, (state) => {
            state.active.status = StatusModel.ERROR;
        })


        .addCase(addLike.pending, (state) => {
            state.likeStatus = StatusModel.LOADING;
        })
        .addCase(addLike.fulfilled, (state, action: PayloadAction<number>) => {
            if(state.active.post) state.active.post.likes = action.payload;
            state.likeStatus = StatusModel.IDLE;
        })
        .addCase(addLike.rejected, (state) => {
            state.likeStatus = StatusModel.ERROR;
        })


        .addCase(removeLike.pending, (state) => {
            state.likeStatus = StatusModel.LOADING;
        })
        .addCase(removeLike.fulfilled, (state, action: PayloadAction<number>) => {
            if (state.active.post) state.active.post.likes = action.payload;
            state.likeStatus = StatusModel.IDLE;
        })
        .addCase(removeLike.rejected, (state) => {
            state.likeStatus = StatusModel.ERROR;
        })


        .addCase(addComment.pending, (state) => {
            state.commentStatus = StatusModel.LOADING;
        })
        .addCase(addComment.fulfilled, (state, action: PayloadAction <CommentType []>) => {
            if (state.active.post) state.active.post.comments = action.payload;
            state.commentStatus = StatusModel.IDLE;
        })
        .addCase(addComment.rejected, (state) => {
            state.commentStatus = StatusModel.ERROR;
        })


        .addCase(removeComment.pending, (state) => {
            state.commentStatus = StatusModel.LOADING;
        })
        .addCase(removeComment.fulfilled, (state, action: PayloadAction <{post: PostType, status: string}>) => {
            if (state.active.post) state.active.post = action.payload.post;
            state.commentStatus = StatusModel.IDLE;
        })
        .addCase(removeComment.rejected, (state) => {
            state.commentStatus = StatusModel.ERROR;
        })

        .addDefaultCase(() => {})
    }
});

const { actions, reducer } = onePostSlice;
export default reducer;
export const { clearActivePost } = actions;