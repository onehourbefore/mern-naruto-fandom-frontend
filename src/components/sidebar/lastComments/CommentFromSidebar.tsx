import React from 'react';
import { useNavigate } from 'react-router-dom';

import { useAppDispatch } from '../../../store';
import { formatDate } from '../../../utils/formatDate';
import { hostName } from '../../../api/apiData';
import { fetchOnePost } from '../../../store/slices/onePostSlice';
import { CommentFromSidebarProps } from '../../../models/CommentsModel';

import cl from './LastCommentsComponent.module.scss';


const CommentFromSidebar: React.FC <CommentFromSidebarProps> = (
    { comment }
): JSX.Element => {
    const navigate = useNavigate();
    const dispatch = useAppDispatch();

    function getPostByComment(id: string): void {
        navigate(`/news/${id}`);
        dispatch(fetchOnePost (id));
    }

    return (
        <div
            data-testid="commentFromSidebar-test"
            className={cl.root__comment}
            onClick={() => getPostByComment(comment.postID)}
        >
            <div className={cl.root__comment_avatar}>
                <img 
                    className={cl.root__comment_avatarIcon} 
                    src={`${hostName}/avatars/${comment.avatar}`} 
                    alt="Аватар" />
            </div>
            <div className={cl.root__comment_content}>
                <div className={cl.root__comment_header}>
                    <div>
                        <h3 className={cl.root__comment_header_title}>
                            {comment.name}
                        </h3>
                        <div className={cl.root__comment_header_email}>
                            {comment.email}
                        </div>
                    </div>
                    <div>
                        <div className={cl.root__comment_header_date}>
                            <div>{formatDate (comment.date)[0]}</div>
                            <div>{formatDate (comment.date)[1]}</div>
                        </div>
                    </div>
                </div>
                <div className={cl.root__comment_body}>
                    {comment.body.length < 40 
                        ? comment.body
                        : `${comment.body.slice (0, 41)}...`}
                </div>
            </div>
        </div>
    )
};

export default CommentFromSidebar;