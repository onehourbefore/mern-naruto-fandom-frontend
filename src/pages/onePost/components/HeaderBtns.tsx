import React from 'react';
import { Link, useNavigate } from 'react-router-dom';
import { useAppDispatch } from '../../../store';
import { StatusModel } from '../../../models/StatusModel';
import { TasksEnum } from '../../../models/EditFormModel';
import { setEditForm } from '../../../store/slices/editFormSlice';
import { clearCreated, clearUpdated } from '../../../store/slices/postsSlice';

import backIcon from '../../../assets/png/back.png';
import editIcon from '../../../assets/png/edit.png';

import cl from '../OnePost.module.scss';


type HeaderBtnsProps = {
    postID: string | undefined,
    authStatus: StatusModel,
    userRole: string,
};

const HeaderBtns: React.FC <HeaderBtnsProps> = (
    {
        postID,
        authStatus,
        userRole,
    }
): JSX.Element => {
    const navigate = useNavigate();
    const dispatch = useAppDispatch();

    function handleSetEditForm() {
        if(postID) {
            dispatch(clearCreated());
            dispatch(clearUpdated());
            dispatch(setEditForm({
                postID: postID, 
                task: TasksEnum.UPDATE,
            }))
            navigate ('/editform');
        }
    }

    return (
        <div
            data-testid='headerBtns'
            className={cl.root__header}
        >
            <Link to="/news">
                <img
                    className={cl.root__header_back}
                    src={backIcon}
                    alt="Назад к ленте постов" />
            </Link>
            {authStatus === StatusModel.SUCCESS && userRole === 'admin' && 
                <img 
                    className={cl.root__header_edit} 
                    src={editIcon} 
                    alt="Редактировать"
                    onClick={handleSetEditForm} />}
        </div>
    );
};

export default HeaderBtns;