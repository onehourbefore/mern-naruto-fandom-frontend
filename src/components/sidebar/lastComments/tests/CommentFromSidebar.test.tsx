import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MemoryRouter } from 'react-router-dom';
import * as reduxHooks from '../../../../store';
import * as actions from '../../../../store/slices/onePostSlice';
import { hostName } from '../../../../api/apiData';
import { StatusModel } from '../../../../models/StatusModel';
import { commentForTest as comment } from './dataForTests';
import CommentFromSidebar from '../CommentFromSidebar';
import HeaderBtns from '../../../../pages/onePost/components/HeaderBtns';



const mockedUseAppDispatch = jest.spyOn(reduxHooks, 'useAppDispatch');
const mockedFetchOnePostAction = jest.spyOn(actions, 'fetchOnePost');


describe('CommentFromSidebar', () => {
    it('first render with comment data', () => {
        mockedUseAppDispatch.mockReturnValue(jest.fn());
        const component = render(
            <MemoryRouter>
                <CommentFromSidebar comment={comment}/>
            </MemoryRouter>
        );
        const avatar = screen.getByAltText(/аватар/i);
        expect(avatar).toBeInTheDocument();
        expect(avatar).toHaveAttribute('src', `${hostName}/avatars/jakeAvatar.img`);
        expect(screen.getByText('Jake')).toBeInTheDocument();
        expect(screen.getByText('jake@gmail.com')).toBeInTheDocument();
        expect(screen.queryByText(comment.body)).toBeNull();
        expect(screen.getByText(`${comment.body.slice(0, 41)}...`)).toBeInTheDocument();
        expect(screen.getByText('10 Nov 2022')).toBeInTheDocument();
        expect(screen.getByText('07:07')).toBeInTheDocument();

        expect(component).toMatchSnapshot();
    });

    it('click to comment', () => {
        mockedUseAppDispatch.mockReturnValue(jest.fn());
        const component = render(
            <MemoryRouter>
                <CommentFromSidebar comment={comment}/>
                <HeaderBtns
                    authStatus={StatusModel.SUCCESS}
                    postID={comment.postID}
                    userRole="admin" />
            </MemoryRouter>
        );

        const commentDiv = screen.getByTestId('commentFromSidebar-test');
        expect(mockedFetchOnePostAction).toHaveBeenCalledTimes(0);
        userEvent.click(commentDiv);
        expect(mockedFetchOnePostAction).toHaveBeenCalledTimes(1);
        expect(mockedFetchOnePostAction).toHaveBeenCalledWith(comment.postID);
        expect(screen.getByAltText(/назад/i)).toBeInTheDocument();
        expect(screen.getByAltText(/редактировать/i)).toBeInTheDocument();
        
        expect(component).toMatchSnapshot();
    });
});
