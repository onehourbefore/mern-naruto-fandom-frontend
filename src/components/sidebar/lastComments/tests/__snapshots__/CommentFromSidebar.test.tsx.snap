// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentFromSidebar click to comment 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="root__comment"
        data-testid="commentFromSidebar-test"
      >
        <div
          class="root__comment_avatar"
        >
          <img
            alt="Аватар"
            class="root__comment_avatarIcon"
            src="http://localhost:5000/avatars/jakeAvatar.img"
          />
        </div>
        <div
          class="root__comment_content"
        >
          <div
            class="root__comment_header"
          >
            <div>
              <h3
                class="root__comment_header_title"
              >
                Jake
              </h3>
              <div
                class="root__comment_header_email"
              >
                jake@gmail.com
              </div>
            </div>
            <div>
              <div
                class="root__comment_header_date"
              >
                <div>
                  10 Nov 2022
                </div>
                <div>
                  07:07
                </div>
              </div>
            </div>
          </div>
          <div
            class="root__comment_body"
          >
            Lorem ipsum dolor sit amet consectetur ad...
          </div>
        </div>
      </div>
      <div
        class="root__header"
        data-testid="headerBtns"
      >
        <a
          href="/news"
        >
          <img
            alt="Назад к ленте постов"
            class="root__header_back"
            src="back.png"
          />
        </a>
        <img
          alt="Редактировать"
          class="root__header_edit"
          src="edit.png"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="root__comment"
      data-testid="commentFromSidebar-test"
    >
      <div
        class="root__comment_avatar"
      >
        <img
          alt="Аватар"
          class="root__comment_avatarIcon"
          src="http://localhost:5000/avatars/jakeAvatar.img"
        />
      </div>
      <div
        class="root__comment_content"
      >
        <div
          class="root__comment_header"
        >
          <div>
            <h3
              class="root__comment_header_title"
            >
              Jake
            </h3>
            <div
              class="root__comment_header_email"
            >
              jake@gmail.com
            </div>
          </div>
          <div>
            <div
              class="root__comment_header_date"
            >
              <div>
                10 Nov 2022
              </div>
              <div>
                07:07
              </div>
            </div>
          </div>
        </div>
        <div
          class="root__comment_body"
        >
          Lorem ipsum dolor sit amet consectetur ad...
        </div>
      </div>
    </div>
    <div
      class="root__header"
      data-testid="headerBtns"
    >
      <a
        href="/news"
      >
        <img
          alt="Назад к ленте постов"
          class="root__header_back"
          src="back.png"
        />
      </a>
      <img
        alt="Редактировать"
        class="root__header_edit"
        src="edit.png"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`CommentFromSidebar first render with comment data 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="root__comment"
        data-testid="commentFromSidebar-test"
      >
        <div
          class="root__comment_avatar"
        >
          <img
            alt="Аватар"
            class="root__comment_avatarIcon"
            src="http://localhost:5000/avatars/jakeAvatar.img"
          />
        </div>
        <div
          class="root__comment_content"
        >
          <div
            class="root__comment_header"
          >
            <div>
              <h3
                class="root__comment_header_title"
              >
                Jake
              </h3>
              <div
                class="root__comment_header_email"
              >
                jake@gmail.com
              </div>
            </div>
            <div>
              <div
                class="root__comment_header_date"
              >
                <div>
                  10 Nov 2022
                </div>
                <div>
                  07:07
                </div>
              </div>
            </div>
          </div>
          <div
            class="root__comment_body"
          >
            Lorem ipsum dolor sit amet consectetur ad...
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="root__comment"
      data-testid="commentFromSidebar-test"
    >
      <div
        class="root__comment_avatar"
      >
        <img
          alt="Аватар"
          class="root__comment_avatarIcon"
          src="http://localhost:5000/avatars/jakeAvatar.img"
        />
      </div>
      <div
        class="root__comment_content"
      >
        <div
          class="root__comment_header"
        >
          <div>
            <h3
              class="root__comment_header_title"
            >
              Jake
            </h3>
            <div
              class="root__comment_header_email"
            >
              jake@gmail.com
            </div>
          </div>
          <div>
            <div
              class="root__comment_header_date"
            >
              <div>
                10 Nov 2022
              </div>
              <div>
                07:07
              </div>
            </div>
          </div>
        </div>
        <div
          class="root__comment_body"
        >
          Lorem ipsum dolor sit amet consectetur ad...
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
